enhancements/authentication/allowing-uri-scheme-in-oidc-sub-claims.md

```diff
@@ -0,0 +1,167 @@
+
+Due to the API limitations mentioned in [Problem Summary](#problem-summary), a
+generic `sub` claim cannot be safely used in order to create an `Identity` object.
+It is apparent that these will need to be encoded in a certain way. Simply encoding
+a `sub` claim won't do as that would break authentication for any OIDC user
+that had its `Identity` created prior to the version containing this fix.
+
+Proposal: if a `sub` claim of an ID token contains either `:` or `/`, url-encode
+the value of the claim and create the new identity object like so:
+
+```yaml
+apiVersion: user.openshift.io/v1
+kind: Identity
+metadata:
+  name: <identity provider name>:url:<url-encoded sub claim value>
+providerName: <identity provider name>
+providerUserName: <url-encoded sub claim value>
+user:
+  name: <preferred username || url-encoded sub claim value>
+  uid: <as usual>
+```
+
+Notice the naming scheme change where the `Identity` object name now consists of
+three sections instead of previous two. The additional section simply contains
+`url` to mark the encoding scheme used in the third section of the name.
+
+The username of the user object also needs to be an encoded `sub` claim value if
+a preferred username is not configured for the identity provider.
+
+### User Stories
+
+#### Story 1
+
+I am a user of an OpenID Connect identity provider and I would like to log in to
+OpenShift.
+
+### Implementation Details
+
+#### oauth-server
+
+When creating a user and an identity object, the oauth-server will have to check
+whether the username it retrieved from the third-party authentication logic contains
+either of `:` or `/`, and if it does, follow the naming scheme from [Proposal](#proposal),
+that is, url-encode the username for `User` objects for the object name, and
+url-encode the username and prepend it with `url:` to use as a username in the
+`Identity` object naming scheme.
+
+#### oauth-apiserver
+
+Validation of the names of the `Identity` objects need to be loosened in order
+to allow three sections separated by the `:` character, where the only allowed
+value for the second section is `url`.
+
+### Risks and Mitigations
+
+When a cluster with existing `Identity` objects of the new naming schemes is downgraded,
+logging in with OIDC providers using URI scheme is their ID tokens' `sub` claims
+will be broken again.
+
+It is also possible that some OpenShift users already started using their OIDC
+providers that were broken due to this, with the use of `unsupportedConfigOverrides`.
+While we generally do not care about these cases, it is fair to say that the
+proposed schema should allow for a scripted conversion of `Identity` and `User`
+objects for them.
+
+## Design Details
+
+### Test Plan
+
+Configure an OIDC provider that uses URL scheme for the `sub` claim in its
+ID tokens (like Keycloak in some configurations), set it up as an OIDC identity
+provider for an OpenShift cluster, and attempt to log in as a user of this
+OIDC provider.
```

Comment:
in automated test or just once manually?

Reply:
I think this could be automated